android/cli/commands/_build.js

```diff
@@ -1267,6 +1267,12 @@ AndroidBuilder.prototype.validate = function validate(logger, config, cli) {
 
 	// determine the abis to support
 	this.abis = this.validABIs;
+	if (this.deployType === 'production') {
+		// by default, remove 'x86' from production builds
+		// 'x86' devices are scarce; this is predominantly used for emulators
+		// we can save 16MB+ by removing this from release builds
+		this.abis.splice(this.abis.indexOf('x86'), 1);
```

Comment:
I could see this being a useful default, but certainly would need some way for users to be able to tune this to explicitly produce a given set of ABIs they wanted to target.

Also, would there be any benefit (in build time, primarily) to a user tuning this to *only* do x86 on development builds when just testing on an x86 emulator?

Reply:
Users can specify the abis they want using the `<abi>` tag (which overrides the removal of x86 from production builds):
```XML
<ti:app xmlns:ti="http://ti.appcelerator.org">
  <android>
    <abi>armeabi-v7a,arm64-v8a</abi>
  </android>
</ti:app>
```

Only doing x86 when targeting emulators wouldn't yield much benefit, maybe a couple of seconds build time saved? But we can't be sure what architecture the emulator is (you can create ARM emulators)

Reply:
Here's an alternative idea.

Our "tiapp.xml" file supports a `<abi/>` entry which allows the dev to select which architectures to include. By default, it includes all architectures supported...
[https://docs.appcelerator.com/platform/latest/#!/guide/tiapp.xml](https://docs.appcelerator.com/platform/latest/#!/guide/tiapp.xml_and_timodule.xml_Reference-section-src-29004921_tiapp.xmlandtimodule.xmlReference-abi)

So, if someone wants the old behavior and exclude the ARM64 architecture, they can add this to their "tiapp.xml" file.
```xml
<ti:app xmlns:ti="http://ti.appcelerator.org">
	<android>
		<abi>armeabi-v7a,x86</abi>
	</android>
</ti:app>
```

What I propose we do is add a `deployType` attribute to the `<abi>` XML element so that a dev can indicate which architectures per deployment type (ie: "test", "development", and "production"). The idea being you can have multiple `<abi>` elements that differ by deployment type, and if the current deployment type `<abi>` cannot be found, then it defaults to the `<abi>` that does not have the deployment type defined... and if that can't be found that it supports all architectures.

For example...
```xml
<ti:app xmlns:ti="http://ti.appcelerator.org">
	<android>
		<abi>armeabi-v7a,arm64-v8a,x86</abi>
		<abi deployType="production">armeabi-v7a,arm64-v8a</abi>
	</android>
</ti:app>
```

Odds are a dev would only define 1 `<abi>` element set to "production" and have all test builds include all architectures, but in the future we might want to add a `split="true"` attribute that the dev can use to generate multiple APKs per architecture.

Also, you can get away with only including ARMv7 architecture for production builds because:
- Most x86 atom devices can run ARMv7 apps via "libhoudini" emulation.
- ARM64 devices can run ARMv7 apps.

x86 is really only useful for fast Android emulator support. ARM64 gives you a slight performance boost, but isn't required (but Google Play may require this architecture to be feature in the future; we'll see).

Thoughts?

Reply:
Thinking as a ti dev that dealt with this a few weeks ago: I‘d love to have a command or option to export each abi as a single apk, bumping the version code for each apk to avoid google play errors. I actually wanted to write a node
script for that but would rather want to see it in the SDK. If we‘d support gradle, it could be done automatically as well.

Reply:
@hansemannn, yeah, I'm still pondering how to approach this. Especially after speaking with a community member here...
https://jira.appcelerator.org/browse/AC-5591

While inconvenient from a test/maintenance standpoint for the app developer, I can understand the desire of wanting to split APKs per architecture to reduce download size for end-users while still supporting best performance for the end-user's device architecture.

Now I'm thinking that we should allow multiple `<abi>` tags of type "production" and build an APK for each one found. And also support a `versionCodeOffset=""` attribute since the version code of a universal APK needs to be different than the per-architecture APKs. This is what Google recommends to do in gradle, although if you look at it, the version code offsetting is kind of kludge on Google's end.

Reply:
That sounds like a great idea!